### ISSUE TITLE
Sorting out Indicators for real-time where getValue(x0) doesn't change with added bars

### ISSUE THREAD

**Issue by jamestomk** [ta4j/ta4j#522]
Is there any documentation on which indicators are relevant to real-time (a useful getValue("now")) versus indicators that may give clear signals only when additional time data is applied?

For example i was stumbling over why ParabolicSarIndicator value(x0) changes over time; i.e., I get a clear signal of trend-change and useful StopLoss price only after some future OHLC data is applied to the underlying time series.   Obviously, I cannot use P-sar for setting a real-time Stop-Loss price.
![cap1](https://user-images.githubusercontent.com/7227076/74349977-0a795600-4d83-11ea-909a-56cc244ecf06.png)
![cap2](https://user-images.githubusercontent.com/7227076/74349989-0ea57380-4d83-11ea-9d2f-cfe2d017ea23.png)
[Example.java.txt](https://github.com/ta4j/ta4j/files/4193177/Example.java.txt)

Why I could use a summary of which Indicators do not change getValue(x0) on additional bar data... This took too long for me to figure out why I was getting discrepancies when I set up fine-tuned candidates of large-run, multi-threaded backtesting that was pre-loading a days-worth of timeseries Bars, and then iterating over each minute.

**Comment by team172011** [comment]
Hi @jamestomk,

I do not get your point. All indicators should not change their calculated value once it has been calculated for any bar. Most indicators are extending the `CachedIndicator` that does exactly this thing: storing the value for `getValue(x0)` so it must only be calculated once.

I can only imagine that you are facing some issues with 'warm up' period. Some indicators need a sufficient amount of input data until they can provide correct values.

Please provide a working example of your set up code if you are assume that there is a bug for any indicator.

**Comment by jamestomk** [comment]
I attach an example with a 1000 bar warmup period that then only looks back starting at PSar #1000 for any changes.
`
 differenct at bar#1000: 49.40162 vs 49.35162
 differenct at bar#1001: 49.67130 vs 49.62130
 differenct at bar#1002: 49.93704 vs 49.88704
 differenct at bar#1003: 50.19963 vs 50.14963
 differenct at bar#1004: 50.45971 vs 50.40970
 differenct at bar#1005: 50.71776 vs 50.66776`

[PSarDriftExample.java.txt](https://github.com/ta4j/ta4j/files/4303717/PSarDriftExample.java.txt)

**Comment by jamestomk** [comment]
I see my java.txt above had at least one project dependency, so instead I (sloppily) merged into your junit test:
[ParabolicSarIndicatorTest.java.txt](https://github.com/ta4j/ta4j/files/4303767/ParabolicSarIndicatorTest.java.txt)

**Comment by team172011** [comment]
I cannot compile your example with the current release (`master `branch) nor with the current SNAPSHOT version (`develop` branch). Can you please try to reproduce your issue with the current version of `develop` branch?

**Comment by jamestomk** [comment]
and better still, if I convert the existing JUnit test 'startUpAndDownTrendTest()' from using MockBars to BaseBar with actual times, the test also fails:
`bars.add(new MockBar(0, 75.9, 76.030000, 74.640000,numFunction));`
vs:
`bars.add(new BaseBar( endTime, (double)0, (double)75.9, (double)76.030000, (double)74.640000, 0, numFunction));   calTA4j.add( Calendar.MINUTE, 1 );`

[ParabolicSarIndicatorTest.java.txt](https://github.com/ta4j/ta4j/files/4303842/ParabolicSarIndicatorTest.java.txt)

**Comment by jamestomk** [comment]
okay just say your last comment, cloned latest "develop" branch and updated my JUnit test
[ParabolicSarIndicatorTest.java.txt](https://github.com/ta4j/ta4j/files/4303851/ParabolicSarIndicatorTest.java.txt)

**Comment by team172011** [comment]
In your file you are still importing classes like `org.ta4j.core.TimeSeries`. Those classes have been renamed in `develop` as well as in `master` branch. That means you are not using the latest version of ta4j. I will only support you with issues that are related to the lates release (0.13) or to the current version (0.14-SNAPSHOT).

Please have a look to the [wiki](https://ta4j.github.io/ta4j-wiki) especially to the [release notes](https://ta4j.github.io/ta4j-wiki/Release-notes.html) further information

**Comment by jamestomk** [comment]
It appears that I didn't update the .txt version, sorry.   reloading
[ParabolicSarIndicatorTest.java.txt](https://github.com/ta4j/ta4j/files/4304017/ParabolicSarIndicatorTest.java.txt)

**Comment by team172011** [comment]
Hi @jamestomk,
thanks for your example. I think this is a bug related to the caching mechanism of `RecursivedCachedIndicator` that is extended by `ParabolicSarIndicator`.

I think your example can also be reproduces by this short unit test (assertion is commented out):

```java
    @Test
    public void testValueInDifferentIndicators() {
    	BarSeries ta4jTimeseries = new BaseBarSeriesBuilder().withNumTypeOf(numFunction).build();
		ParabolicSarIndicator pSarInd = new ParabolicSarIndicator(ta4jTimeseries);
		Random r = new Random(0);
		float price = 45.5f;
		float priceIncr = 0.25f;
		for ( ; price < 60.0f; price+= priceIncr ) {
			if ( r.nextInt( 100 ) > 85 ) { 
				priceIncr = 0-priceIncr;
			}
			ZonedDateTime endTime = ZonedDateTime.now().plusMinutes(counter++);
			ta4jTimeseries.addBar(endTime, numFunction.apply(price), numFunction.apply(price + 0.5f), numFunction.apply(price - 0.5f), numFunction.apply(price), numFunction.apply(1000), numFunction.apply(0));
			Num tempResult = new ParabolicSarIndicator(ta4jTimeseries).getValue(ta4jTimeseries.getEndIndex());
			Num existingResult = pSarInd.getValue(ta4jTimeseries.getEndIndex());
			// result pSarInd with this bar series should be the same as the result of a newly created indicator with this bar series
			if(!tempResult.isEqual(existingResult)) {
				
				//just to be sure..
				assert(pSarInd.getBarSeries().getBarCount() == ta4jTimeseries.getBarCount());
				assert(pSarInd.getBarSeries().getLastBar().equals(ta4jTimeseries.getLastBar()));
				
				System.out.println(String.format("Difference at bar %s existing: %s temp: %s", ta4jTimeseries.getEndIndex(), tempResult.toString(), existingResult.toString()));
				// this will fail for all:
				// assert(tempResult.isEqual(existingResult));
			}
		}
    }
```

```java
It can also be reproduced by just using one indicator instance and comparing the results in the end:

    @Test
    public void testSameValueInGrowingSeries() {
    	List<Num> growingResultCache = new ArrayList<>();
    	BarSeries ta4jTimeseries = new BaseBarSeriesBuilder().withNumTypeOf(numFunction).build();
		ParabolicSarIndicator pSarInd = new ParabolicSarIndicator(ta4jTimeseries);
		Random r = new Random(0);
		float price = 45.5f;
		float priceIncr = 0.25f;
		for ( ; price < 60.0f; price+= priceIncr ) {
			if ( r.nextInt( 100 ) > 85 ) { 
				priceIncr = 0-priceIncr;
			}
			ZonedDateTime endTime = ZonedDateTime.now().plusMinutes(counter++);
			ta4jTimeseries.addBar(endTime, numFunction.apply(price), numFunction.apply(price + 0.5f), numFunction.apply(price - 0.5f), numFunction.apply(price), numFunction.apply(1000), numFunction.apply(0));
			growingResultCache.add(pSarInd.getValue(ta4jTimeseries.getEndIndex()));
		}
		System.out.println(ta4jTimeseries.getBarCount() +" "+ pSarInd.getBarSeries().getBarCount());
		for(int i = ta4jTimeseries.getBeginIndex(); i <= ta4jTimeseries.getEndIndex(); i++) {
			System.out.println("Compare cache: " + growingResultCache.get(i) + " vs. getValue:" + pSarInd.getValue(i));
			// this will fail for all:
			assert(growingResultCache.get(i).isEqual(pSarInd.getValue(i)));
		}
    }
```
A quick fix (that may heavily slow down performance) for this would be to not extend from the `RecursiveCachedIndicator`, so that all results will be calculated again:

```java
public class ParabolicSarIndicator extends AbstractIndicator<Num> {

    private final Num maxAcceleration;
    private final Num accelerationIncrement;
    private final Num accelerationStart;
    private Num accelerationFactor;
    private boolean currentTrend; // true if uptrend, false otherwise
    private int startTrendIndex = 0; // index of start bar of the current trend
    private LowPriceIndicator lowPriceIndicator;
    private HighPriceIndicator highPriceIndicator;
    private Num currentExtremePoint; // the extreme point of the current calculation
    private Num minMaxExtremePoint; // depending on trend the maximum or minimum extreme point value of trend

    /**
     * Constructor with default parameters
     *
     * @param series the bar series for this indicator
     */
    public ParabolicSarIndicator(BarSeries series) {
        this(series, series.numOf(0.02), series.numOf(0.2), series.numOf(0.02));

    }

    /**
     * Constructor with custom parameters and default increment value
     *
     * @param series the bar series for this indicator
     * @param aF     acceleration factor
     * @param maxA   maximum acceleration
     */
    public ParabolicSarIndicator(BarSeries series, Num aF, Num maxA) {
        this(series, aF, maxA, series.numOf(0.02));
    }

    /**
     * Constructor with custom parameters
     *
     * @param series    the bar series for this indicator
     * @param aF        acceleration factor
     * @param maxA      maximum acceleration
     * @param increment the increment step
     */
    public ParabolicSarIndicator(BarSeries series, Num aF, Num maxA, Num increment) {
        super(series);
        highPriceIndicator = new HighPriceIndicator(series);
        lowPriceIndicator = new LowPriceIndicator(series);
        maxAcceleration = maxA;
        accelerationFactor = aF;
        accelerationIncrement = increment;
        accelerationStart = aF;
    }

    @Override
    public Num getValue(int index) {
        Num sar = NaN;
        if (index == getBarSeries().getBeginIndex()) {
            return sar; // no trend detection possible for the first value
        } else if (index == getBarSeries().getBeginIndex() + 1) {// start trend detection
            currentTrend = getBarSeries().getBar(getBarSeries().getBeginIndex()).getClosePrice()
                    .isLessThan(getBarSeries().getBar(index).getClosePrice());
            if (!currentTrend) { // down trend
                sar = highPriceIndicator.getValue(index); // put sar on high price of candlestick
                currentExtremePoint = sar;
                minMaxExtremePoint = currentExtremePoint;
            } else { // up trend
                sar = lowPriceIndicator.getValue(index); // put sar on low price of candlestick
                currentExtremePoint = sar;
                minMaxExtremePoint = currentExtremePoint;

            }
            return sar;
        }

        Num priorSar = getValue(index - 1);
        if (currentTrend) { // if up trend
            sar = priorSar.plus(accelerationFactor.multipliedBy((currentExtremePoint.minus(priorSar))));
            currentTrend = lowPriceIndicator.getValue(index).isGreaterThan(sar);
            if (!currentTrend) { // check if sar touches the low price
                sar = minMaxExtremePoint; // sar starts at the highest extreme point of previous up trend
                currentTrend = false; // switch to down trend and reset values
                startTrendIndex = index;
                accelerationFactor = accelerationStart;
                currentExtremePoint = getBarSeries().getBar(index).getLowPrice(); // put point on max
                minMaxExtremePoint = currentExtremePoint;
            } else { // up trend is going on
                currentExtremePoint = new HighestValueIndicator(highPriceIndicator, index - startTrendIndex)
                        .getValue(index);
                if (currentExtremePoint.isGreaterThan(minMaxExtremePoint)) {
                    incrementAcceleration();
                    minMaxExtremePoint = currentExtremePoint;
                }

            }
        } else { // downtrend
            sar = priorSar.minus(accelerationFactor.multipliedBy(((priorSar.minus(currentExtremePoint)))));
            currentTrend = highPriceIndicator.getValue(index).isGreaterThanOrEqual(sar);
            if (currentTrend) { // check if switch to up trend
                sar = minMaxExtremePoint; // sar starts at the lowest extreme point of previous down trend
                accelerationFactor = accelerationStart;
                startTrendIndex = index;
                currentExtremePoint = getBarSeries().getBar(index).getHighPrice();
                minMaxExtremePoint = currentExtremePoint;
            } else { // down trend io going on
                currentExtremePoint = new LowestValueIndicator(lowPriceIndicator, index - startTrendIndex)
                        .getValue(index);
                if (currentExtremePoint.isLessThan(minMaxExtremePoint)) {
                    incrementAcceleration();
                    minMaxExtremePoint = currentExtremePoint;
                }
            }
        }
        return sar;
    }

    /**
     * Increments the acceleration factor.
     */
    private void incrementAcceleration() {
        if (accelerationFactor.isGreaterThanOrEqual(maxAcceleration)) {
            accelerationFactor = maxAcceleration;
        } else {
            accelerationFactor = accelerationFactor.plus(accelerationIncrement);
        }
    }
}
```

But I think we should try to fix the caching logic here.

**Comment by jamestomk** [comment]
Does this mean that any other classes based on 'RecursiveCachedIndicator' (e.g., 'EMAIndicator') could also have errors?
   If so, I might volunteer to add to JUnit test cases for these, based on your feedback on how to simplify my own JUnit test case.   Not that I may be the most eloquent code swashbuckler as you may have noticed!

Of less importance.... I know you're busy!
If I try this version based on AbstractIndicator<> (and change from 'calculate()' to 'getValue()', I seem to get the same failures in my unit test, so I must be overlooking something.

**Comment by nimo23** [comment]
>If so, I might volunteer to add to JUnit test cases for these

We need junit test cases for those sensible parts of the `RecursiveCachedIndicator` instead.

**And yes, I did this suggestion few years ago here and do it again:** Improving the caching layer! 
- provide test cases for those sensible parts of `RecursiveCachedIndicator`. 
- make the caching layer optional: The user must have the option **to opt out** from the caching layer (can be a property within `RecursiveCachedIndicator`: 'disableCaching'). With this, comparisions between "cached indicator values" and "non cached indicator" values can be easily made (if there are any deviations, then it is obvious that the caching layer has some bugs or need improvements..)
- make the RECURSION_THRESHOLD a property
- and other small improvements which can be made..?

I hope @team172011 are now willing (after few years) that my opinions about those was and is reasonable.

**Comment by andrei-baraian** [comment]
Hello,

I am also facing problems with the real time parabolic SAR indicator.
First thing I noticed is that if you call getValue(series.endIndex) multiple times, the indicator starts giving high/low values sequentially instead of the actual SAR value. This can be reproduced by modifying slightly the existing unit test like this (re-wrote it in Kotlin):

```
   @ Test
   fun startUpAndDownTrendTest() {
        val numFunction: Function<Number, Num> = Function { t: Number -> DoubleNum.valueOf(t) }
        var date = Instant.parse("2019-11-28T00:00:00.000Z")
        val bars = arrayListOf<Bar>()
        bars.add(MockBar(Duration.ofSeconds(5), ZonedDateTime.ofInstant(date.plusSeconds(5), ZoneId.of("Z")),74.1, 75.1, 74.06, 75.11, 34405.1,460.0, 4, numFunction))
        bars.add(MockBar(Duration.ofSeconds(5), ZonedDateTime.ofInstant(date.plusSeconds(10), ZoneId.of("Z")),74.7, 75.9, 76.030000, 74.640000,34711.0,460.0, 4, numFunction))
        bars.add(MockBar(Duration.ofSeconds(5), ZonedDateTime.ofInstant(date.plusSeconds(15), ZoneId.of("Z")),75.7, 75.24, 76.269900, 75.060000,34721.293,460.0, 4, numFunction))
        bars.add(MockBar(Duration.ofSeconds(5), ZonedDateTime.ofInstant(date.plusSeconds(20), ZoneId.of("Z")),74.7, 75.17, 75.280000, 74.500000,34371.4,460.0,4, numFunction))

        val sar = ParabolicSarIndicator(MockBarSeries(bars));

        assertEquals("NaN", sar.getValue(0).toString());
        assertNumEquals(74.640000000000000568434188608080, sar.getValue(1))
        assertNumEquals(74.640000000000000568434188608080, sar.getValue(2)) // start with up trend
        assertNumEquals(76.269900000000006912159733474255, sar.getValue(3)) // switch to downtrend

        assertNumEquals(76.269900000000006912159733474255, sar.getValue(3)) // switch to downtrend
    }
```


I can see the the following potential issues in the indicator class:
1. accelerationFactor is incremented at each call of the calculate() function if the trend didn't change, making the SAR value increase/decrease wrongly. I think there should be a previousAccelerationFactor used and it's value saved with the bar index.
2. The initial currentExtremePoint and minMaxExtremePoint is set to the first value of SAR, while I think these should be initialized with the opposite extreme, like:
```
if (!currentTrend) { // down trend
    sar = highPriceIndicator.getValue(index); // put sar on max price of candlestick
    currentExtremePoint = lowPriceIndicator.getValue(index);
    minMaxExtremePoint = currentExtremePoint;
} else { // up trend
    sar = lowPriceIndicator.getValue(index); // put sar on min price of candlestick
    currentExtremePoint = highPriceIndicator.getValue(index);
    minMaxExtremePoint = currentExtremePoint;
}
```

**Comment by gant85** [comment]
Hi, do you have any update for this problem?

**Comment by jamestomk** [comment]
> Hi, do you have any update for this problem?

While I thought I had time to contribute, I must now bow out on any commitment to a fix.  I have worked around it in my own use.

**Comment by team172011** [comment]
This bug was not related to the recursive cache indicator but to a bug in the ParabolicSarIndicator #902 and has been fixed with #903
